### PR TITLE
Feature/expected categories

### PR DESCRIPTION
**Pull request by ColdTeapot273K**
Add support for processing only explicitly expected categories for `preprocessing.OneHotEncoder`, `preprocessing.OrdinalEncoder`, akin to `sklearn` api for respective encoders.

All doctests pass (i've added some).

Rationale:
`sklearn` has this neat feature where you can explicitly pass category values you want to see in the encoder state, other values are filtered out. See `categories` parameter: [OneHotEncoder](https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.OneHotEncoder.html), [OrdinalEncoder](https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.OrdinalEncoder.html)

This is convenient when you work with high cardinality category spaces where some values are rare and you want to regularize your model. E.g. I've had a practical problem where constraining only to pre-selected top 20% frequent categories in 1 000 000 cardinality space can give you a 10%+ latency boost with no significant loss in metrics, and also make a model lighter on RAM.

This implementation is hackable so if user wants to modify lists of expected categories between training steps, they can do so by direct attribute access. E.g. can glue with modules like [TargetAgg](https://riverml.xyz/latest/api/feature-extraction/TargetAgg/) for some cool dynamic reevaluation of expected category lists. 

P.S. Pls bump Ruff, my LSP config compains coz api changes. Also MyPy complained a lot about about `str | dict | defaultdict` type hints for `category` parameter, I just had to give up on them, maybe someone has better ideas how to  handle them.

### CODE REVIEW

**Comment by MaxHalford** [comment]
> P.S. Pls bump Ruff, my LSP config compains coz api changes. Also MyPy complained a lot about about str | dict | defaultdict type hints for category parameter, I just had to give up on them, maybe someone has better ideas how to handle them.

Duly noted, I'll take a look.

File: river/preprocessing/one_hot.py [hunk 1]
```diff
@@ -85,6 +85,29 @@ class OneHotEncoder(base.MiniBatchTransformer):
     {'c2_h': 1}
     {'c2_e': 1}
 
+    Like in `scikit-learn`, you can also specify the expected categories manually.
+    This is handy when you want to constrain category encoding space
+    to e.g. top 20% most popular category values you've picked in advance.
+
+    >>> categories = {'c1': {'a', 'h'}, 'c2': {'x', 'e'}}
+    >>> oh = preprocessing.OneHotEncoder(categories=categories)
+    >>> # oh = preprocessing.OneHotEncoder()
```
Comment (MaxHalford) [inline river/preprocessing/one_hot.py:94]:
```suggestion
```

File: river/preprocessing/one_hot.py [hunk 2]
```diff
@@ -192,23 +215,43 @@ class OneHotEncoder(base.MiniBatchTransformer):
...
+
+    # oh = preprocessing.OneHotEncoder()
+    >>> oh.learn_many(X)
+    >>> df = oh.transform_many(X)
+    >>> df.sort_index(axis="columns")
+       c1_a  c1_h  c2_e  c2_x
+    0     0     0     0     0
+    1     1     0     0     1
+    2     0     0     0     0
     """
```
Comment (MaxHalford) [inline river/preprocessing/one_hot.py:230]:
```suggestion
    2     0     0     0     0
    
    """
```

**Review by MaxHalford — COMMENTED** [review]
Thanks for this PR, it's useful.

My preference here would be to not adhere to sklearn, and set `categories` to `None` rather than `auto`.

**Comment by ColdTeapot273K** [comment]
> Thanks for this PR, it's useful.
> 
> My preference here would be to not adhere to sklearn, and set `categories` to `None` rather than `auto`.

Deal. I shall modify.

**Comment by MaxHalford** [comment]
@ColdTeapot273K sorry for not replying in a while! I like the changes, we can merge. Before that though, could you add an entry to [`unreleased.md`](https://github.com/online-ml/river/blob/main/docs/releases/unreleased.md)?

**Comment by ColdTeapot273K** [comment]
@MaxHalford no problem, i understand.

Done, please check.
